core/src/main/java/io/camunda/migrator/RuntimeMigrator.java

```diff
@@ -339,4 +365,25 @@ public void setMode(MigratorMode mode) {
     this.mode = mode;
   }
 
+  private Map<String, Long> getC8DeploymentIdToKeyMap() {
+    int from = 0;
+    List<ProcessDefinition> c8Deployments;
+    Map<String, Long> idToKey = new HashMap<>();
+
+    do {
+      final int finalFrom = from;
+      c8Deployments = camundaClient.newProcessDefinitionSearchRequest().page(p -> {
+        p.limit(batchSize);
+        p.from(finalFrom);
+      }).execute().items();
+
+      idToKey.putAll(c8Deployments.stream()
+          .collect(
+              Collectors.toMap(ProcessDefinition::getProcessDefinitionId, ProcessDefinition::getProcessDefinitionKey)));
+      from = from + c8Deployments.size();
+    } while (c8Deployments.size() == batchSize);
+
+    return idToKey;
+  }
```

Comment:
/suggest
Instead of querying for all process definitions all the time, can we just query for the process definition that we actually need on validation? There may be more process definitions deployed on C8 than on C7, but we only need the C8 process definition key for C7 legacy process instances.

See code suggestions above.